storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +257,182 @@ func (s *commitLogSource) Read(
-			for start, unmergedBlock := range unmergedBlocks.encoders {
-				block := blocksPool.Get()
-				if len(unmergedBlock) == 0 {
-					emptyErrs++
-					continue
-				} else if len(unmergedBlock) == 1 {
-					block.Reset(start, unmergedBlock[0].enc.Discard())
-				} else {
-					readers := make([]io.Reader, len(unmergedBlock))
-					for i := range unmergedBlock {
-						stream := unmergedBlock[i].enc.Stream()
-						readers[i] = stream
-						defer stream.Finalize()
-					}
+		mergeSemaphore <- true
+		wg.Add(1)
+		go func(shard int, unmergedShard map[ts.Hash]encodersByTime) {
+			var shardResult result.ShardResult
+			shardResult, shardEmptyErrs[shard], shardErrs[shard] = s.mergeShard(
+				unmergedShard, blocksPool, multiReaderIteratorPool, encoderPool, blopts)
+			if len(shardResult.AllSeries()) > 0 {
+				// Prevent race conditions while updating bootstrapResult from multiple go-routines
+				bootstrapResultLock.Lock()
+				// Shard is a slice index so conversion to uint32 is safe
+				bootstrapResult.Add(uint32(shard), shardResult, nil)
+				bootstrapResultLock.Unlock()
+			}
+			<-mergeSemaphore
+			wg.Done()
+		}(shard, unmergedShard)
+	}
 
-					iter := multiReaderIteratorPool.Get()
-					iter.Reset(readers)
-
-					var err error
-					enc := encoderPool.Get()
-					enc.Reset(start, blopts.DatabaseBlockAllocSize())
-					for iter.Next() {
-						dp, unit, annotation := iter.Current()
-						encodeErr := enc.Encode(dp, unit, annotation)
-						if encodeErr != nil {
-							err = encodeErr
-							errs++
-							break
-						}
-					}
-					if iterErr := iter.Err(); iterErr != nil {
-						if err == nil {
-							err = iter.Err()
-						}
-						errs++
-					}
+	// Wait for all merge goroutines to complete
+	wg.Wait()
+	s.printMergeShardsOutcome(shardErrs, shardEmptyErrs)
+	return bootstrapResult
+}
 
-					iter.Close()
-					for i := range unmergedBlock {
-						unmergedBlock[i].enc.Close()
-					}
+func (s *commitLogSource) mergeShard(
+	unmergedShard map[ts.Hash]encodersByTime,
+	blocksPool block.DatabaseBlockPool,
+	multiReaderIteratorPool encoding.MultiReaderIteratorPool,
+	encoderPool encoding.EncoderPool,
+	blopts block.Options,
+) (shardResult result.ShardResult, numShardEmptyErrs int, numErrs int) {
+	shardResult = result.NewShardResult(len(unmergedShard), s.opts.ResultOptions())
+	for _, unmergedBlocks := range unmergedShard {
+		blocks := block.NewDatabaseSeriesBlocks(len(unmergedBlocks.encoders))
+		for start, unmergedBlock := range unmergedBlocks.encoders {
+			startInNano := time.Unix(0, start)
+			block := blocksPool.Get()
+			if len(unmergedBlock) == 0 {
+				numShardEmptyErrs++
+				continue
+			} else if len(unmergedBlock) == 1 {
+				block.Reset(startInNano, unmergedBlock[0].enc.Discard())
+			} else {
+				readers := make([]io.Reader, len(unmergedBlock))
+				for i := range unmergedBlock {
+					stream := unmergedBlock[i].enc.Stream()
+					readers[i] = stream
+					defer stream.Finalize()
```

Comment:
Should not perform defer inside a huge for loop for large dynamic allocation of deferred fn calls and also just not finalizing for so long.

Reply:
Yeah I think I tried removing this at one point and was surprised that it didn't seem to have any performance impact. It was probably done this way originally because when you put the streams in the readers slice you lose access to the Finalize() method, but I was able to move it to the end manually by doing a type assertion. Let me know if it looks like I'm finalizing it at the right place, because its not happening earlier than it used to (by virtue of doing it in the middle of the loop instead of at the very end when the function returns)